### PR TITLE
bpo-40025: Require _generate_next_value_ to be defined before members

### PR DESCRIPTION
**Pull request by abbyonstott**
https://bugs.python.org/issue40025

### CODE REVIEW

File: Lib/enum.py [hunk 1]
```diff
@@ -77,6 +77,9 @@ def __setitem__(self, key, value):
                     ):
                 raise ValueError('_names_ are reserved for future Enum use')
             if key == '_generate_next_value_':
+                # check if members already defined
+                if len(self._member_names) > len(dir(super)):
```
Comment (ethanfurman) [inline Lib/enum.py]: `if self._member_names` is sufficient -- if any have been it will evaluate as *truthy*

**Review by ethanfurman — CHANGES_REQUESTED** [review]
Please add a note in the docs about `_generate_next_value_` needing to be defined before any members are.

**Comment by tirkarthi** [comment]
Could this also have a test about the order raising TypeError?

**Comment by ethanonstott** [comment]
I have made the requested changes; please review again

File: Misc/NEWS.d/next/Library/2020-03-21-05-26-38.bpo-40025.DTLtyq.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Throw error when _generate_next_value_ is defined after members. Patch by Ethan Onstott.
```
Comment (ethanfurman) [inline Misc/NEWS.d/next/Library/2020-03-21-05-26-38.bpo-40025.DTLtyq.rst]: "Throw error" --> "Raise TypeError"

File: Lib/enum.py [hunk 3]
```diff
@@ -77,6 +77,9 @@ def __setitem__(self, key, value):
                     ):
                 raise ValueError('_names_ are reserved for future Enum use')
             if key == '_generate_next_value_':
+                # check if members already defined
+                if self._member_names:
+                    raise TypeError("_generate_next_value_ must be defined before members")
```
Comment (ethanfurman) [inline Lib/enum.py:83]: Looks like the change broke the test at line #1104.   Maybe a new `auto_called` flag that starts as `False` and changes to `True` when `auto()` is used the first time, and this test changes to `if auto_called:`.

**Review by ethanfurman — CHANGES_REQUESTED** [review]
Getting closer -- thank you for persevering!

**Comment by ethanonstott** [comment]
I have made the requested changes; please review again

File: Lib/enum.py (hunk 3 above)
Comment (abbyonstott) [reply Lib/enum.py:83]: I added `auto_called` as a instance attribute of `_EnumDict`. Is that what I should have done in this situation?

File: Lib/enum.py [hunk 4]
```diff
@@ -100,6 +104,9 @@ def __setitem__(self, key, value):
                 # enum overwriting a descriptor?
                 raise TypeError('%r already defined as: %r' % (key, self[key]))
             if isinstance(value, auto):
+                # on first call to auto set auto_called to True
+                if not self.auto_called:
```
Comment (ethanfurman) [inline Lib/enum.py]: Skip the `if` and just always set `auto_called` to True.

File: Lib/enum.py [hunk 5]
```diff
@@ -60,6 +60,7 @@ def __init__(self):
         self._member_names = []
         self._last_values = []
         self._ignore = []
+        self.auto_called = False
```
Comment (ethanfurman) [inline Lib/enum.py]: Let's use an `_` prefix:  `self._auto_called`.

**Review by ethanfurman — CHANGES_REQUESTED** [review]
Looking good!  A couple minor changes and we should be good to go.

**Comment by ethanonstott** [comment]
I have made the requested changes; please review again

File: Lib/enum.py (hunk 3 above)
Comment (ethanfurman) [reply Lib/enum.py:83]: Yup.  :)

**Review by ethanfurman — APPROVED** [review]

**Pull request merged**